### DISCUSSION TITLE
TuyaMCU support for Moes shutter/blinds: buttons link to dpid with payload

### DISCUSSION THREAD

**Discussion by Dobie007** [arendst/Tasmota#13308] [Show and tell]
![screen](https://user-images.githubusercontent.com/59510536/136656605-a986bcb9-b140-4249-95a5-25624908a1a7.png)
I have got a set of curtain switches that originally had a Beken7231 MCU which I replaced with a Tasmota flashed TYWE2S ESP8285. Only the 3.3V power and the serial interface is connected so any ESP8266 will do the job as well.
![screen2](https://user-images.githubusercontent.com/59510536/136656367-9b4af17c-26bc-4ced-ba3d-ed38f194fc60.png)

I configured Tasmota in "Shuttermode 1" and as "Tuya MCU 54" and got the three buttons to be fired if i send: 
TuyaSend4 1,0;  TuyaSend4 1,1;  TuyaSend4 1,2 via console.
I mapped 3 relays to three unsed dpid's using:  TuyaMCU 11,2 TuyaMCU 12,3 TuyaMCU 13,4

To control the module i would need:
fnid11 <-> dpid 1 payload 0 
fnid12 <-> dpid 1 payload 1
fnid13 <-> dpid 1 payload 2

It would be good to be able to define something like "TuyaMCU 12,1 1" or so... 

Here the actual messages the module sends back when the keys were pressed:

Left button:
19:04:11.911 {"TuyaReceived":{"Data":"55AA00070005010400010213","Cmnd":7,"CmndData":"0104000102","DpType4Id1":2,"1":{"DpId":1,"DpIdType":4,"DpIdData":"02"}}}

Center button:
19:20:43.813 {"TuyaReceived":{"Data":"55AA00070005010400010112","Cmnd":7,"CmndData":"0104000101","DpType4Id1":1,"1":{"DpId":1,"DpIdType":4,"DpIdData":"01"}}}

right button:
19:21:41.632 {"TuyaReceived":{"Data":"55AA00070005010400010011","Cmnd":7,"CmndData":"0104000100","DpType4Id1":0,"1":{"DpId":1,"DpIdType":4,"DpIdData":"00"}}}

**Comment by Dobie007** [comment]
I managed to solve the issue using the Rules:

First I configured the module as shutter see https://tasmota.github.io/docs/Blinds-and-Shutters
Then I configured it as follows:
TuyaMCU":[{"fnId":11,"dpId":2},{"fnId":12,"dpId":3},{"fnId":13,"dpId":4}]
Setting the buttons to unused dpid's

which shows up as:
![grafik](https://user-images.githubusercontent.com/59510536/136694238-756c4d6c-a96a-4bbc-8324-39ca93ed3900.png)

Rule1 
ON Power1#state=1 DO TuyaSend4 1,2 ENDON 
ON Power2#state=1 DO TuyaSend4 1,0 ENDON 
ON Power3#state=1 DO TuyaSend4 1,1 ENDON 
ON Power1#state=0 DO TuyaSend4 1,1 ENDON 
ON Power2#state=0 DO TuyaSend4 1,1 ENDON 
ON Power3#state=0 DO TuyaSend4 1,1 ENDON

Activated the rule with "Rule1 1"

and finally checked the MQTT outputs to be working in all ocasions.

To get the LED'S on the switches to light up I had to switch dpid 7 on.

Tasmota is fascinating - thanks guys!

**Reply by mikip4** [reply]
Hello, problem is en version of tasmota.
Install 9.5.0

then

put your IP address:
IPAddress1 192.168.0.xx
IPAddress2 192.168.0.1
IPAddress3 255.255.255.0
IPAddress4 8.8.8.8
restart 1

in config/ other put:
{"NAME":"Moes WS-EUR-CW","GPIO":[0,2272,0,2304,0,0,0,0,0,0,0,0,0,0],"FLAG":0,"BASE":54}

console:
SetOption80 1
ShutterMode 1
Interlock 1,2
Interlock ON

TuyaMCU 11,2
TuyaMCU 12,3
TuyaMCU 13,4

Rule1 ON Power1#state=1 DO TuyaSend4 1,0 ENDON ON Power2#state=1 DO TuyaSend4 1,2 ENDON ON Power3#state=1 DO TuyaSend4 1,1 ENDON ON Power1#state=0 DO TuyaSend4 1,1 ENDON ON Power2#state=0 DO TuyaSend4 1,1 ENDON ON Power3#state=0 DO TuyaSend4 1,1 ENDON Rule1 1

ShutterOpenDuration DURATION -----------    OPEN OR CLOSE IN SECONDS

PowerOnState 0                     ---------Relays OFF on start

Thanks a lot!!

**Comment by Dobie007** [comment]
Update: In order to get the shutter position working correctly, even when operating the roler with the keys i needed to update the rules. The problem behind was that the esp was not recognizing when a key is pressed. Here the refined rules:

Rule1 
ON Power1#state=1 DO TuyaSend4 1,2 ENDON 
ON TuyaReceived#Data=55AA00070005010400010213 DO Power1 1 BREAK 
ON Power1#state=0 DO TuyaSend4 1,1 ENDON

ON Power2#state=1 DO TuyaSend4 1,0 ENDON 
ON TuyaReceived#Data=55AA00070005010400010011 DO Power2 1 BREAK
ON Power2#state=0 DO TuyaSend4 1,1 ENDON

ON Power3#state=1 DO TuyaSend4 1,1 ENDON 
ON TuyaReceived#Data=55AA00070005010400010112 DO Power3 TOGGLE ENDON

The only flaw I have now is that once a button is pressed it refires itself until the stop button is pressed - but that is no problem.

**Comment by DNJ1980** [comment]
Hi. Sorry for my english was translated with google. I have the same switch only I soldered an ESP-02S into it. I configured the switch as a shutter and took the template that Tasmota has but I only have one button... I tried your plan with Tuya MCU 54 but 3 buttons just don't want to come up. what am I doing wrong? Please help

**Reply by Dobie007** [reply]
![grafik](https://user-images.githubusercontent.com/59510536/152675318-98b1c001-c843-4137-831f-1b4bc33cca75.png)
This is what your module config should look  like.
Set the logging to debug: 
![grafik](https://user-images.githubusercontent.com/59510536/152675508-c7526f6f-333a-4227-adf2-364df09e6405.png)
then you should see the heartbeat of the Tuya chip already in the console:
![grafik](https://user-images.githubusercontent.com/59510536/152675549-98560f33-ac83-4876-a46e-d188fa80a12d.png)
If not you should check your soldering with a microscope...

If that works send these commands to the TUYA chip by console: 
fnid11 <-> dpid 1 payload 0
Your button should fire then: LED and relay on.
If that works you have to go on with the rules..

**Comment by DNJ1980** [comment]
Hello, I did everything so it works as you showed me. It only happens when I enter fnid11 <-> dpid 1 payload 0.
Result= Command unknown.

Edith: if i have TuyaSend4 1.0 i can control it. 1.0, 1.1, 1.2

**Comment by Dobie007** [comment]
Check out this page: https://tasmota.github.io/docs/TuyaMCU/

**Comment by DNJ1980** [comment]
I'm a complete novice with some technical understanding. I could use some help with creating a rule 🙂

**Reply by Dobie007** [reply]
The thing you seem to be missing is the mapping of the TUYA MCU:  
First define the functionid and dpid by these commands:
TuyaMCU 11,2 TuyaMCU 12,3 TuyaMCU 13,4  That creates the buttons and matches them to unused dpid's.

Then:
It sends the info which button is pressed on one dpid with the different payloads 0,1 and 2 depending on the button you pressed.
So now you need to tell Tasmota how to handle the input by using a rule (several rules in one, to be precise)
As given above: 
Example:

ON TuyaReceived#Data=55AA00070005010400010213 DO Power1 1 BREAK

This means: If the Tuya MCU sends the given data String: "55AA00070005010400010213" 
interpret that as "Power1 = 1" 
Break means forget the following rules for now.

You need to put everything from above what I posted under "Rule 1" in one command: 
 Read this to see how to append rules...  https://tasmota.github.io/docs/Rules/#rule-command
I just reformatted it into several sentences so it is easier to understand.

**Comment by DNJ1980** [comment]
![Unbenannt](https://user-images.githubusercontent.com/60576062/152689015-9d863289-3a78-497f-8107-daed91dbf18b.jpg)

I can do what I want... unfortunately it doesn't work... Is it maybe because I don't have enough switches?

**Reply by Dobie007** [reply]
Its quite some time since i did it myself...
Did you send these three commands via "Konsole":

TuyaMCU 11,2 
TuyaMCU 12,3 
TuyaMCU 13,4

Those should generate the extra buttons.

**Comment by DNJ1980** [comment]
Hi. I got it. I changed the rule a bit because it didn't work properly for me. Thank you for your help and patience

**Comment by mikip4** [comment]
Hello,
when I created new buttons there are not push buttons , and images are not like arrows.
Sorry for my bad english.
Any solutions?
Thanks a lot in advance.

Miquel

**Reply by chrimez** [reply]
> First I configured the module as shutter see https://tasmota.github.io/docs/Blinds-and-Shutters 

I assume you failed to correctly activate ShutterMode. 
It also failed for me, so I followed [this](https://blakadder.com/tuya-covers/#tasmota-setup): 

> To control the blinds from Tasmota webUI we will map dpId 2 as a dimmer with TuyaMCU 21,2 and control them with the dimmer slider or using Dimmer commands. [...]

**Comment by chrimez** [comment]
Thanks @Dobie007 this helped me a lot jointly with [this reference](https://blakadder.com/tuya-covers/).

I am only struggling: how to calibrate, i.e. change the definition of end-points of my shade motor? 
- _**TuyaSend2 2,100**_  will move it to existing endpoint ("fully closed", "100%")
- yet, the shade is still half open, so I need to change this endpoint further out
- also, I currently don't know how to move the motor beyond current endpoints. As all "TuyaMCU" commands will stop the motor at current end points.

Any hints appreciated. Thanks

**Reply by Dobie007** [reply]
Maybe this helps:  https://www.youtube.com/watch?v=Z-grrvnu2bU

**Comment by chrimez** [comment]
OK thanks. My understanding is: 
The linked video is about how to use **ESP8266 controlled** motors, based on [Tasmota command set](https://tasmota.github.io/docs/Commands/#shutters).
However, my device uses **TuyaMCU controlled** motor relays, bound to the [TuyaMCU command set](https://tasmota.github.io/docs/Commands/#tuyamcu): TuyaMCU, TuyaSend<n>

I tried the Tasmota standard commands: shuttermode  shutterbutton1  shuttercalibration1 .
For me none of them do anything:  {"Command":"Unknown"}

**Reply by Dobie007** [reply]
Tasmota always runs on an ESP826x (or  ESP32).  
The shuttermode is only a specific way to handle the buttons. The TuyaMCu is the interface to the buttons - with a specific Tuya serial protocol.

So you need to to both: Configure it as Shutter and the set the interface to Tuya. The further steps are described above.
You need to tell Tasmota what to send when a button is pressed.

**Reply by Dobie007** [reply]
"Shuttermode 1" is probably what you are missing...  That gives you the up and down buttons.

This is the console output if you have it right:
18:04:33.701 CMD: Shuttermode
18:04:33.708 MQT: stat/Rollo/RESULT = {"ShutterMode":1}

**Comment by chrimez** [comment]
Thanks again.
On Tasmota, seems ShutterMode does not work for me. Maybe it's not compiled in my build - not sure. It's the OEM build shipped by manufacturer ("Shut It" shades actor by wesmartify).

Actually, Tasmota buttons an commands are not needed to move the shutters. I assume you had your calibration done / endpoints set correctly before starting this thread.

**Reply by benjaminnestler** [reply]
Did you solve your issue @chrimez ?
